src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,120 @@
+
+        private static readonly LocalizableString s_localizableTitleQualifyMembers = new LocalizableResourceString(nameof(FeaturesResources.AddThisOrMeQualification), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly DiagnosticDescriptor s_descriptorQualifyMemberAccess = new DiagnosticDescriptor(IDEDiagnosticIds.AddQualificationDiagnosticId,
+                                                                    s_localizableTitleQualifyMembers,
+                                                                    s_localizableMessage,
+                                                                    DiagnosticCategory.Style,
+                                                                    DiagnosticSeverity.Hidden,
+                                                                    isEnabledByDefault: true,
+                                                                    customTags: DiagnosticCustomTags.Unnecessary);
+
+        private OptionSet _lazyDefaultOptionSet;
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_descriptorQualifyMemberAccess);
+
+        protected abstract bool IsAlreadyQualifiedMemberAccess(SyntaxNode node);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.FieldReferenceExpression, OperationKind.PropertyReferenceExpression, OperationKind.MethodReferenceExpression);
+        }
+
+        private void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            var memberReference = (IMemberReferenceExpression)context.Operation;
+
+            // this is a static reference so we don't care if it's qualified
+            if (memberReference.Instance == null)
+            {
+                return;
+            }
+
+            // if we're not referencing `this.` or `Me.` (e.g., a parameter, local, etc.)
+            if (memberReference.Instance.Kind != OperationKind.InstanceReferenceExpression)
+            {
+                return;
+            }
+
+            // if there's already a `this.` or `Me.` qualification
+            if (IsAlreadyQualifiedMemberAccess(memberReference.Instance.Syntax))
+            {
+                return;
+            }
+
+            var optionSet = GetOptionSet(context.Options);
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var memberSymbol = GetSymbol(memberReference);
+            if (memberSymbol == null)
+            {
+                return;
+            }
+
+            // if the reference is to a base type
+            if (memberSymbol.ContainingSymbol != memberReference.Instance.ResultType)
```

Comment:
This seems odd. Why would I not want `this.SomeMethodInByBase()`? And in general, do we have enough tests to make sure no sort of `base` references break anything here?